### PR TITLE
Pretrain

### PR DESCRIPTION
**Pull request by rrevoid**

### CODE REVIEW

File: scripts/train_model.py [hunk 1]
```diff
@@ -23,13 +24,18 @@ def configure_arg_parser():
...
     parser.add_argument("-nw", "--num_workers", type=int, default=1, help="Number of workers for dataloader")
-    parser.add_argument("-sc", "--schedule", default="cosine", help="Noise schedule for diffusion diffusion")
+    parser.add_argument("-sc", "--schedule", default="sqrt", help="Noise schedule for diffusion diffusion")
     parser.add_argument("-df", "--diffusion_steps", type=int, default=2000, help="Number of diffusion steps")
     parser.add_argument("-s", "--num_steps", type=int, default=100000, help="Number of training steps")
     parser.add_argument("-l", "--logging_step", type=int, default=100, help="Logging step")
     parser.add_argument(
         "-vi", "--val_interval", type=int, default=10000, help="Number of training steps between evaluations"
     )
+    parser.add_argument("-p", "--pretrain", action=argparse.BooleanOptionalAction, help="Flag for model pretraining")
```
Comment (SpirinEgor) [inline scripts/train_model.py]: `action="store_true"` -- recommended way by documentation

File: src/data/dataset.py [hunk 2]
```diff
@@ -101,3 +103,59 @@ def collate_fn(self, samples: list[Dialog], return_all_candidates: bool = True):
...
+
+        logger.info(f"Loading dataset from '{path}'")
+
+        with open(path, "r") as f:
+            lines_list = list(f)
+
+        for line in lines_list:
+            result = json.loads(line)
+            dialog = {"context": [result["src"]], "candidates": [result["trg"]]}
+            self.dataset.append(Dialog(**dialog))
```
Comment (SpirinEgor) [inline src/data/dataset.py]: Do you have enough memory to load the entire data set? We usually use an iterative dataset that loads the dataset row by row on request.

File: src/data/dataset.py [hunk 3]
```diff
@@ -101,3 +103,59 @@ def collate_fn(self, samples: list[Dialog], return_all_candidates: bool = True):
...
+
+        logger.info(f"Loaded {len(self.dataset)} samples")
+
+    def __len__(self):
+        return len(self.dataset)
+
+    def __getitem__(self, idx: int) -> Dialog:
+        return self.dataset[idx]
+
+    def collate_fn(self, samples: list[Dialog]):
```
Comment (SpirinEgor) [inline src/data/dataset.py]: It appears that you are copying the matching function from the ConvAI dataset, but this is not necessary. We don't have any candidates for the Commonsense dataset, so things are much simpler

```python
def __getitem__(self, idx: int) -> tuple[str, str]:
        return self.dataset[idx]```

def collate_fn(self, samples: list[tuple[str, str]]):
    str_contexts, str_candidates = zip(*samples)

    # [batch size, context seq len]
    b_contexts = self.context_tokenizer(
            str_contexts,
            max_length=self.max_context_len,
            padding=True,
            truncation=True,
            return_tensors="pt",
            add_special_tokens=False,
        ).input_ids

        # Tokenizer truncates on the left, but for candidates we want to truncate on the right
        b_candidates = self.candidate_tokenizer(
            str_candidates, padding="max_length", return_tensors="pt", add_special_tokens=False
        ).input_ids

        return b_contexts, b_candidates
```

File: src/diffusion/model.py [hunk 4]
```diff
@@ -11,10 +12,26 @@
...
+        return model.encoder, model.decoder, emb_dim
+
+    elif mode == "bert":
+        decoder_config = BertConfig(
+            vocab_size=model.config.vocab_size,
+            is_decoder=True,
+            hidden_size=emb_dim,
+            num_attention_heads=num_attention_heads,
+            add_cross_attention=True,
+            cross_attention_hidden_size=emb_dim,
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: It seems that this parameter never used :(

File: src/diffusion/model.py [hunk 5]
```diff
@@ -11,10 +12,26 @@
...
-def get_components(name):
+def get_components(name: str, mode: str = "same", num_attention_heads: int = 16):
     model = AutoModel.from_pretrained(name)
+    emb_dim = model.config.d_model
+
+    if mode == "same":
+        return model.encoder, model.decoder, emb_dim
+
+    elif mode == "bert":
+        decoder_config = BertConfig(
```
Comment (SpirinEgor) [inline src/diffusion/model.py:28]: We also should consider these parameters:
- `num_hidden_layers` (default 12)
- `num_attention_heads` (default 12)
- `intermediate_size` (default 3072)

For example, we can use 3-layer transformer with 2560 feedforward dim. Also, I suggest to use the same number of attention heads as in blenderbot. It doesn't affect number of parameters

File: src/data/dataset.py [hunk 6]
```diff
@@ -148,14 +146,14 @@ def collate_fn(self, samples: list[Dialog]):
...
+        # Tokenizer truncates on the left, but for replies we want to truncate on the right
+        b_replies = self.reply_tokenizer(
+            str_replies, padding="max_length", return_tensors="pt", add_special_tokens=False
         ).input_ids
-        b_candidates = b_candidates[:, : self.max_target_len]
-        # [batch size, # candidates, candidates seq len]
-        b_candidates = b_candidates.view(len(samples), -1, b_candidates.size(1))
+        b_replies = b_replies[:, : self.max_target_len]
+        # [batch size, # replies, replies seq len]
+        b_replies = b_replies.view(len(samples), -1, b_replies.size(1))
```
Comment (SpirinEgor) [inline src/data/dataset.py]: Do we really need second dimension?

File: src/data/dataset.py [hunk 7]
```diff
@@ -101,3 +103,57 @@ def collate_fn(self, samples: list[Dialog], return_all_candidates: bool = True):
...
+
+        # Tokenizer truncates on the left, but for replies we want to truncate on the right
+        b_replies = self.reply_tokenizer(
+            str_replies, padding="max_length", return_tensors="pt", add_special_tokens=False
+        ).input_ids
+        b_replies = b_replies[:, : self.max_target_len]
+        # [batch size, # replies, replies seq len]
+        b_replies = b_replies.view(len(samples), -1, b_replies.size(1))
+
+        return b_contexts, b_replies.squeeze(1)
```
Comment (SpirinEgor) [inline src/data/dataset.py]: Why do you reshape your tensor and then squeeze it back?

**Review by SpirinEgor — APPROVED** [review]
